crates/prune/src/segments/history.rs

```diff
@@ -34,79 +33,180 @@ where
+            PruneStepResult::ReachedDeletedEntriesLimit |
+            PruneStepResult::ReachedBlockLimit => break,
+            PruneStepResult::MaybeMoreData => processed += 1,
+        }
+    }
+
+    Ok((processed, limiter.deleted_entries_count()))
+}
+
+/// Steps once with given cursor.
+///
+/// Caution! Prune job timeout is not checked, only limit on deleted entries count. This allows for
+/// a clean exit of a prune job that's pruning different tables concurrently, by letting them step
+/// to the same height before timing out.
+pub(crate) fn step_prune_indices<DB, T, SK>(
+    cursor: &mut <<DB as Database>::TXMut as DbTxMut>::CursorMut<T>,
+    limiter: &mut PruneLimiter,
+    key_matches: &impl Fn(&T::Key, &T::Key) -> bool,
+    last_key: &impl Fn(&T::Key) -> T::Key,
+) -> Result<PruneStepResult, DatabaseError>
+where
+    DB: Database,
+    T: Table<Value = BlockNumberList>,
+    T::Key: AsRef<ShardedKey<SK>>,
+{
+    trace!(target: "pruner::history", "pruning history");
+
+    let Some((key, blocks)) = cursor.next()? else {
+        trace!(target: "pruner::history",
+            "cursor is at end of search, nothing more to prune",
+        );
+
+        return Ok(PruneStepResult::Finished)
+    };
+
+    let Some(to_block) = limiter.last_pruned_block() else {
+        trace!(target: "pruner::history",
+            "missing target block"
+        );
+
+        return Ok(PruneStepResult::ReachedBlockLimit)
+    };
+
+    if let Some(limit) = limiter.deleted_entries_limit() {
+        if limiter.deleted_entries_count() == limit {
+            trace!(target: "pruner::history",
+                "reached limit on deleted entries"
+            );
+
+            return Ok(PruneStepResult::ReachedDeletedEntriesLimit)
         }
+    }
+
+    // If shard consists only of block numbers less than the target one, delete shard
+    // completely.
+    if key.as_ref().highest_block_number <= to_block {
+        cursor.delete_current()?;
+        limiter.increment_deleted_entries_count();
+
+        trace!(target: "pruner::history",
+            "deleted complete shard"
+        );
+
+        if key.as_ref().highest_block_number == to_block {
+            // Shard contains only block numbers up to the target one, so we can skip to
+            // the last shard for this key. It is guaranteed that further shards for this
+            // sharded key will not contain the target block number, as it's in this shard.
+            cursor.seek_exact(last_key(&key))?;
+        }
+    } else {
+        trace!(target: "pruner::history",
+            to_block,
+            highest_block_shard=key.as_ref().highest_block_number,
+            "cannot prune complete shard"
+        );
+
         // Shard contains block numbers that are higher than the target one, so we need to
         // filter it. It is guaranteed that further shards for this sharded key will not
         // contain the target block number, as it's in this shard.
```

Comment:
```suggestion
        // Shard contains block numbers that are higher than the target one, so we need to
        // filter it. It is guaranteed that further shards for this sharded key will not
        // contain the target block number, if it's in this shard.
```

@shekhirin ? shard may contain block numbers that are only higher than the target one too.

Reply:
> shard may contain block numbers that are only higher than the target one too.

right, true